pkg/portal/sms/service.go

```diff
@@ -0,0 +1,46 @@
+package sms
+
+import (
+	"context"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/lib/infra/sms/smsapi"
+	"github.com/authgear/authgear-server/pkg/lib/infra/sms/twilio"
+	"github.com/authgear/authgear-server/pkg/portal/model"
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+type Logger struct{ *log.Logger }
+
+func NewLogger(lf *log.Factory) Logger {
+	return Logger{lf.New("portal-sms")}
+}
+
+type Service struct {
+	Logger Logger
+}
+
+func (s *Service) SendByTwilio(ctx context.Context,
+	app *model.App,
+	to string,
+	cfg model.SMSProviderConfigurationTwilioInput,
+) error {
+
+	twilioClient := twilio.NewTwilioClient(&config.TwilioCredentials{
+		AccountSID:          cfg.AccountSID,
+		AuthToken:           cfg.AuthToken,
+		MessagingServiceSID: cfg.MessagingServiceSID,
+	})
+
+	translationService := NewTranslationService(app)
+	sender, err := translationService.GetSenderForTestSMS(ctx)
+	if err != nil {
+		return err
+	}
+
+	return twilioClient.Send(ctx, smsapi.SendOptions{
+		Sender: sender,
```

Comment:
I discover a design problem here. We should allow the developer to configure `sms.default.sender` in this page. And in the send-test-sms mutation input, we should take the sender, instead of deriving the sender from the current project configuration.

The SMTP page has the same problem. See https://linear.app/authgear/issue/DEV-2310/allow-setting-default-sender-in-the-custom-email-provider-page

`sms.default.sender` is not always used. We should document this in the configuration page as well. In particular, when Twilio messaging service SID is specified, the sender is ignored.

cc @fungc-io

Reply:
I see, got it

Reply:
I think I will handle this in later pr.